File: 77-combinations/combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let combinations = []

    const backtrack = (start, path) => {
        if (path.length === k) {
            combinations.push([...path])
            return
        }

        for (let i = start; i <= n; i++) {
            path.push(i)
            backtrack(i + 1, path)
            path.pop()
        }
    }

    backtrack(1, [])

    return combinations
};

File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans = []

        def backtrack(curr: List[int], first_num: int) -> None:
            if len(curr) == k:
                ans.append(curr.copy())
                return

            for num in range(first_num, n + 1):
                curr.append(num)

                backtrack(curr, num+1)

                curr.pop()

        backtrack([], 1)
        return ans